### PR TITLE
Add Moment.__sub__ and generalize Moment.__add__

### PR DESCRIPTION
**Pull request by Strilanc**
- `Moment.__add__` now accepts any `OP_TREE`
- `Moment.__sub__` takes an `OP_TREE` and returns a moment with those operations removed
- If there is an operation-to-remove that was not present, an error is raised
- Also bump mypy version to fix an incompatibility with python 3.8 and fix new warnings
    - Add explicit !r to format statements touching bytes
    - Rename `value_equality.py` to avoid name collision with the decorator it defines
    - Ignore false positives related to abstract base type bounds

### CODE REVIEW

File: dev_tools/auto_merge.py [hunk 1]
```diff
@@ -509,7 +508,7 @@ def attempt_sync_with_master(pr: PullRequestDetails
     data = {
         'base': pr.branch_name,
         'head': master_sha,
-        'commit_message': 'Update branch (automerge)'.format(pr.branch_name)
+        'commit_message': 'Update branch (automerge)'
```
Comment (viathor) [inline dev_tools/auto_merge.py:511]: The original looks like it was accidentally missing a {}. Suggestion:

```python
'commit_message': 'Update branch {pr.branch_name} (automerge)'
```

File: dev_tools/auto_merge.py [hunk 2]
```diff
@@ -534,8 +533,8 @@ def attempt_sync_with_master(pr: PullRequestDetails
             "'Update Branch' for me before trying again.")
 
     raise RuntimeError('Sync with master failed for unknown reason. '
-                       'Code: {}. Content: {}.'.format(response.status_code,
-                                                       response.content))
+                       'Code: {}. Content: {!r}.'.format(
+                           response.status_code, response.content))
```
Comment (viathor) [inline dev_tools/auto_merge.py:537]: Use a format string?

(Applies throughout.)

File: dev_tools/auto_merge.py (hunk 1 above)
Comment (Strilanc) [reply dev_tools/auto_merge.py:511]: No, it was intentional. When I changed it to not mention the branch name I just forgot to remove the format call.

File: dev_tools/auto_merge.py (hunk 2 above)
Comment (Strilanc) [reply dev_tools/auto_merge.py:537]: Acknowledged. Wanted to keep the refactoring minimal.

**Pull request merged**